Add caching to reduce object updates

ostromart opened the pull request:
Moved to new repo from https://github.com/istio/operator/pull/667.

ostromart committed:
Add caching to reduce object updates

istio-testing commented:
@ostromart: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 24a12351208ea02d53793a25ac4bc60aebf65050 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/6918) | `/test unit-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

rcernich commented on operator/pkg/helmreconciler/pruning.go:
```diff
@@ -17,38 +17,32 @@ package helmreconciler
+func (h *HelmReconciler) Prune(excluded map[string]bool, all bool) error {
 	allErrors := []error{}
-	namespacedResources, nonNamespacedResources := h.customizer.PruningDetails().GetResourceTypes()
+	namespacedResources, _ := h.customizer.PruningDetails().GetResourceTypes()
```
Is there a reason you're leaving non-namespaced resources around?  We should be cleaning up things like `ClusterRole`, etc.

ostromart replied:
It's to make rollback possible - at least we can't delete CRDs because that will delete user config. Can other the other cluster resources be deleted? I wasn't sure.

rcernich replied:
If CRDs aren't in the type list, you don't have to worry about them being deleted.  If the cluster resource wasn't installed, I think it's safe to assume it's not needed.  If it breaks something, that's an issue with the rendered charts (e.g. deleting a ClusterRole that's used by a RoleBinding or ClusterRoleBinding, etc.).

Also, it's not relevant to this PR, but since you brought it up, CRD installation should be part of a general bootstrapping operation, rather than part of control plane installation.

ostromart replied:
i've removed the CRD from the list and left everything else for removal.

rcernich commented on operator/pkg/helmreconciler/rendering.go:
```diff
@@ -41,6 +42,13 @@ import (
+var (
+	// objectCache holds the latest copy of each object applied by the controller, keyed by its Hash() function.
+	// Must be locked since ProcessManifest can be run concurrently.
+	objectCache   = make(map[string]*object.K8sObject)
```
It seems like the cache should be associated with the ICP vs global.  Also, it doesn't look like objects are ever removed from the cache.

ostromart replied:
even though there's no overlap between objects in different ICPs, the locking goes away, so this is a great idea - not sure why that didn't occur to me before as i was somehow fixated on cache correctness. 
regarding the object removal, is there a time we'd need/want to remove objects from the cache?

rcernich replied:
As long as the cache is cleared at the start of any rendering operation, I don't see a need to clear it.  With the previous implementation, it seemed like the cache would never be cleaned, meaning nothing would be pruned (i.e. removed objects would still be in the cache).

ostromart committed:
Remove cache locking, make gen

rcernich commented on operator/pkg/helmreconciler/rendering.go:
```diff
@@ -209,9 +199,7 @@ func (h *HelmReconciler) ProcessManifest(manifest manifest.Manifest) (int, error
-		objectCacheMu.Lock()
-		objectCache[obj.Hash()] = obj
-		objectCacheMu.Unlock()
+		h.objectCache[obj.Hash()] = obj
```
Because ProcessManifest may be called from multiple threads, doesn't this still need to be protected by a mutex?

ostromart replied:
i don't think so because each thread will have a copy of h. which also makes me realize that i have to go back to the global cache since unfortunately the helm reconciler is created each time we receive the CR update. we could keep it hanging around but that will be even more messy than having a global cache with locking.

rcernich replied:
It probably should be initialized with the original cache, to help in processing of the updates, but should be replaced with new versions when the update is complete, but before going through the prune.

If the global cache accounts for resources in different namespaces, belonging to different ICPs, then global is probably fine.  The key, as mentioned above is to ensure the cache correctly reflects what's actually been installed (i.e. no stale entries).

ostromart replied:
yes, i think there could be a problem with cache not being pruned - objects will be pruned the first time because pruning is based on the manifest list not the cache but the next time around the same object may not be written to the API server if it was pruned the last time because it's still in the cache. 
you're correct that we shouldn't have namespace overlap between CRs, but even if we do it just results in some unnecessary writes to the server, not data consistency problems. so i think we are good here.

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 24a12351208ea02d53793a25ac4bc60aebf65050 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/6918) | `/test unit-tests_istio`
gencheck_istio | a9949a5a049b8a0f75e969234c66756b1ea7a8d9 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5203) | `/test gencheck_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | a9949a5a049b8a0f75e969234c66756b1ea7a8d9 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5203) | `/test gencheck_istio`
unit-tests_istio | a9949a5a049b8a0f75e969234c66756b1ea7a8d9 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/6977) | `/test unit-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ostromart committed:
Restore global cache, prune cluster resources

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | a9949a5a049b8a0f75e969234c66756b1ea7a8d9 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/6977) | `/test unit-tests_istio`
gencheck_istio | 013fad17dedb21b9de4d56ba47c4d97a063f8dfa | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5219) | `/test gencheck_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | 013fad17dedb21b9de4d56ba47c4d97a063f8dfa | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5219) | `/test gencheck_istio`
unit-tests_istio | 013fad17dedb21b9de4d56ba47c4d97a063f8dfa | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/6993) | `/test unit-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ostromart committed:
Rework the cache and locking to be per-CR name

ostromart commented:
@rcernich per our discussion above, i've reworked the cache so that it's partitioned by CR name rather than global. this was needed to be able to prune any removed objects from the per-name cache without removing any from other CRs. 
i added an extra serialization to prevent multiple workers accessing the same CR's cache. this was to  reduce the amount of locking logic (and anyway not sure that it's a good idea to interleave such operations).
ptal since the locking logic has become more tricky than i would like and could use some scrutiny.

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 013fad17dedb21b9de4d56ba47c4d97a063f8dfa | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/6993) | `/test unit-tests_istio`
gencheck_istio | 77f772fe80f24f41727f5033e0f1d01e22a8c348 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5233) | `/test gencheck_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ostromart committed:
Merge branch 'master' of https://github.com/istio/istio into cache-resources

ostromart committed:
Refactor per offline feedback

istio-testing commented:
@ostromart: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | 6bde4dcc32edb0103bdeacda4eee0a96f1d95721 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5246) | `/test gencheck_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | 6bde4dcc32edb0103bdeacda4eee0a96f1d95721 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5246) | `/test gencheck_istio`
unit-tests_istio | 6bde4dcc32edb0103bdeacda4eee0a96f1d95721 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/7021) | `/test unit-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ostromart committed:
Merge branch 'master' of https://github.com/istio/istio into cache-resources

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 6bde4dcc32edb0103bdeacda4eee0a96f1d95721 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/7021) | `/test unit-tests_istio`
gencheck_istio | f88e40eb8c131849a32a8c30148ca51d44090c1e | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5251) | `/test gencheck_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mandarjog commented on operator/pkg/helmreconciler/rendering.go:
```diff
@@ -41,6 +42,19 @@ import (
 
+// ObjectCache is a cache of objects.
+type ObjectCache struct {
+	cache map[string]*object.K8sObject
```
Please indicate the key here.

ostromart replied:
Done.

mandarjog commented on operator/pkg/helmreconciler/rendering.go:
```diff
@@ -175,19 +189,77 @@ func unmarshalAndValidateIOPSpec(iopsYAML string) (*v1alpha1.IstioOperatorSpec,
+		}
+	}
+	objectCache := objectCaches[name]
+
```
`objectCachesMu.Unlock()` here instead of later?

ostromart replied:
Done.

ostromart commented on operator/pkg/helmreconciler/rendering.go:
```diff
@@ -175,19 +189,77 @@ func unmarshalAndValidateIOPSpec(iopsYAML string) (*v1alpha1.IstioOperatorSpec,
+
+	// Create and/or get the cache corresponding to the CR name we're processing. Per name partitioning is required to
+	// prune the cache to remove any objects not in the manifest generated for a given CR.
+	if objectCaches[name] == nil {
```
i may be missing your point, but your version would be:
	objectCache := objectCaches[name]
	if objectCache == nil {
		objectCaches[name] = &ObjectCache{
			cache: make(map[string]*object.K8sObject),
			mu:    &sync.RWMutex{},
		}
        	objectCache = objectCaches[name]
	}
it saves a map access at the cost of an extra line, which is why i wrote it the way i did.

ostromart committed:
Comments

ostromart committed:
Merge branch 'master' of https://github.com/istio/istio into cache-resources

istio-testing commented:
@ostromart: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | 80a04dd5821c0fa3048badd436ba756403dc0229 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5280) | `/test gencheck_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | 80a04dd5821c0fa3048badd436ba756403dc0229 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5280) | `/test gencheck_istio`
integ-istioio-k8s-tests_istio | 80a04dd5821c0fa3048badd436ba756403dc0229 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-istioio-k8s-tests_istio/3674) | `/test integ-istioio-k8s-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ostromart committed:
Make cache per component, add debug

rcernich commented on operator/pkg/helmreconciler/rendering.go:
```diff
@@ -175,19 +190,77 @@ func unmarshalAndValidateIOPSpec(iopsYAML string) (*v1alpha1.IstioOperatorSpec,
+
+	// Ensure that for a given CR name only one control loop uses the per-name cache at any time.
+	objectCache.mu.Lock()
+	defer objectCache.mu.Unlock()
```
If I read this correctly, we're effectively holding onto the cache lock for this entire call, which means the threads invoking `ProcessManifest()` become serialized through this function, negating any benefit to splitting the processing up across multiple threads.

ostromart replied:
objectCachesLock (plural) is global and that's held only briefly to get the objectCache lock, which is per-component, so components are not serialized. what is serialized is if two control loops cannot work on the same component for the same CR (as might happen if a reconcile was still in progress and another update came in). that's something that we should really be doing anyway regardless of caching, since i'm not sure if we can rely on eventual consistency to work in all such cases.

ostromart committed:
Merge branch 'master' of https://github.com/istio/istio into cache-resources

ostromart committed:
Merge branch 'master' of https://github.com/istio/istio into cache-resources

ostromart committed:
Fix resourceVersion update

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-istioio-k8s-tests_istio | 80a04dd5821c0fa3048badd436ba756403dc0229 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-istioio-k8s-tests_istio/3674) | `/test integ-istioio-k8s-tests_istio`
gencheck_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5854) | `/test gencheck_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-istioio-k8s-tests_istio | 80a04dd5821c0fa3048badd436ba756403dc0229 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-istioio-k8s-tests_istio/3674) | `/test integ-istioio-k8s-tests_istio`
gencheck_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5854) | `/test gencheck_istio`
lint_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/lint_istio/7322) | `/test lint_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-istioio-k8s-tests_istio | 80a04dd5821c0fa3048badd436ba756403dc0229 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-istioio-k8s-tests_istio/3674) | `/test integ-istioio-k8s-tests_istio`
gencheck_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5854) | `/test gencheck_istio`
lint_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/lint_istio/7322) | `/test lint_istio`
unit-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/7651) | `/test unit-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-istioio-k8s-tests_istio | 80a04dd5821c0fa3048badd436ba756403dc0229 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-istioio-k8s-tests_istio/3674) | `/test integ-istioio-k8s-tests_istio`
gencheck_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5854) | `/test gencheck_istio`
lint_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/lint_istio/7322) | `/test lint_istio`
unit-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/7651) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-pilot-k8s-tests_istio/7317) | `/test integ-pilot-k8s-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5854) | `/test gencheck_istio`
lint_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/lint_istio/7322) | `/test lint_istio`
unit-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/7651) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-pilot-k8s-tests_istio/7317) | `/test integ-pilot-k8s-tests_istio`
integ-istioio-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-istioio-k8s-tests_istio/4127) | `/test integ-istioio-k8s-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5854) | `/test gencheck_istio`
lint_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/lint_istio/7322) | `/test lint_istio`
unit-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/7651) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-pilot-k8s-tests_istio/7317) | `/test integ-pilot-k8s-tests_istio`
integ-istioio-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-istioio-k8s-tests_istio/4127) | `/test integ-istioio-k8s-tests_istio`
integ-security-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-security-k8s-tests_istio/7489) | `/test integ-security-k8s-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ostromart committed:
Lint, fix controller test

ostromart committed:
Merge branch 'master' of https://github.com/istio/istio into cache-resources

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/lint_istio/7322) | `/test lint_istio`
unit-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/7651) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-pilot-k8s-tests_istio/7317) | `/test integ-pilot-k8s-tests_istio`
integ-istioio-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-istioio-k8s-tests_istio/4127) | `/test integ-istioio-k8s-tests_istio`
integ-security-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-security-k8s-tests_istio/7489) | `/test integ-security-k8s-tests_istio`
gencheck_istio | e90c2d80abe1fdb86f6b34e7a965930aea916ca8 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5858) | `/test gencheck_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/unit-tests_istio/7651) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-pilot-k8s-tests_istio/7317) | `/test integ-pilot-k8s-tests_istio`
integ-istioio-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-istioio-k8s-tests_istio/4127) | `/test integ-istioio-k8s-tests_istio`
integ-security-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-security-k8s-tests_istio/7489) | `/test integ-security-k8s-tests_istio`
gencheck_istio | e90c2d80abe1fdb86f6b34e7a965930aea916ca8 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5858) | `/test gencheck_istio`
lint_istio | e90c2d80abe1fdb86f6b34e7a965930aea916ca8 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/lint_istio/7327) | `/test lint_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ostromart committed:
Lint

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-pilot-k8s-tests_istio/7317) | `/test integ-pilot-k8s-tests_istio`
integ-istioio-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-istioio-k8s-tests_istio/4127) | `/test integ-istioio-k8s-tests_istio`
integ-security-k8s-tests_istio | d51e88652f5768327c260ab3f59f2112a7864b21 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/integ-security-k8s-tests_istio/7489) | `/test integ-security-k8s-tests_istio`
lint_istio | e90c2d80abe1fdb86f6b34e7a965930aea916ca8 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/lint_istio/7327) | `/test lint_istio`
gencheck_istio | faaf0e32185628e7745f3da0bce0f092ea6ba3a1 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/20344/gencheck_istio/5860) | `/test gencheck_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ostromart committed:
make gen

ostromart commented:
@rcernich i just fixed a couple of outstanding issues and everything appears to be working well - could you please take another look?

istio-testing commented:
In response to a cherrypick label: new pull request created: #20720

ostromart commented:
/cherrypick release-1.5

istio-testing commented:
@ostromart: #20344 failed to apply on top of branch "release-1.5":
```Applying: Add caching to reduce object updates
error: Failed to merge in the changes.
Using index info to reconstruct a base tree...
M	operator/pkg/controller/istiocontrolplane/pruningdetails.go
M	operator/pkg/helmreconciler/pruning.go
M	operator/pkg/helmreconciler/reconciler.go
M	operator/pkg/helmreconciler/rendering.go
M	operator/pkg/object/objects.go
M	operator/pkg/object/objects_test.go
Falling back to patching base and 3-way merge...
Auto-merging operator/pkg/helmreconciler/rendering.go
CONFLICT (content): Merge conflict in operator/pkg/helmreconciler/rendering.go
Auto-merging operator/pkg/helmreconciler/pruning.go
CONFLICT (content): Merge conflict in operator/pkg/helmreconciler/pruning.go
Patch failed at 0001 Add caching to reduce object updates

```

<details>

In response to [this](https://github.com/istio/istio/pull/20344#issuecomment-580853594):

>/cherrypick release-1.5


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>
